src/mito2/src/read/merge.rs

```diff
@@ -201,32 +219,143 @@ impl BatchMerger {
     /// reset the buffer.
-    fn merge_batches(&mut self) -> Result<Option<Batch>> {
+    fn merge_batches(&mut self) -> Result<VecDeque<Batch>> {
         if self.batches.is_empty() {
-            return Ok(None);
+            return Ok(VecDeque::new());
         }
 
-        let batches = mem::take(&mut self.batches);
-        // Concat all batches.
-        let mut batch = Batch::concat(batches)?;
-
-        // TODO(yingwen): metrics for sorted and unsorted batches.
-        if !self.is_sorted {
-            // Slow path. We need to merge overlapping batches. For simplicity, we
-            // just sort the all batches and remove duplications.
-            batch.sort_and_dedup()?;
-            // We don't need to remove duplications if timestamps of batches
-            // are not overlapping.
-        }
+        let mut output = VecDeque::with_capacity(self.batches.len());
+        if self.is_sorted {
+            // Fast path. We can output batches directly.
+            for batch in self.batches.drain(..) {
+                output_batch(&mut output, batch)?;
+            }
 
-        // Filter rows by op type. Currently, the reader only removes deleted rows but doesn't filter
-        // rows by sequence for simplicity and performance reason.
-        batch.filter_deleted()?;
+            return Ok(output);
+        }
 
-        // Reset merger.
+        // Slow path. We need to merge overlapping batches.
+        // Constructs a heap from batches. Batches in the heap is not empty, we need to check
+        // this before pushing a batch into the heap.
+        let mut heap = BinaryHeap::from_iter(self.batches.drain(..).map(CompareTimeSeq));
+        // Reset merger as sorted as we have cleared batches.
         self.is_sorted = true;
 
-        Ok(Some(batch))
+        // Sorts batches.
+        while let Some(top) = heap.pop() {
+            let top = top.0;
+            let Some(next) = heap.peek() else {
+                // If there is no remaining batch, we can output the top-most batch.
+                output_batch(&mut output, top)?;
+                break;
+            };
+            let next = &next.0;
+
+            if top.last_timestamp() < next.first_timestamp() {
+                // If the top-most batch doesn't overlaps with the next batch, we can output it.
+                output_batch(&mut output, top)?;
+                continue;
+            }
+
+            // Safety: Batches (top, next) in the heap is not empty, so we can use unwrap here.
+            // Min timestamp in the next batch.
+            let next_min_ts = next.first_timestamp().unwrap();
+            let timestamps = top.timestamps_native().unwrap();
+            // Binary searches the timestamp in the top batch.
+            // Safety: Batches should have the same timestamp resolution so we can compare the native
+            // value directly.
+            match timestamps.binary_search(&next_min_ts.value()) {
```

Comment:
So the top batch is always sorted and deduped, am I right? Or is every batch already sorted and deduped but must be merged, deduped, and sorted?

Reply:
Every batch is already sorted and deduped.